User guided specialisation stage

Oli-Ar opened the pull request:
Draft PR for implementing a user-guided specialisation pass in the mlscript compiler, very much still WIP

LPTK commented on hkmc2/shared/src/test/mlscript/specialisation/Basic.mls:
```diff
@@ -0,0 +1,87 @@
+
+
+:spt
```
It would be helpful to also show the JS code, which is much easier to read.

Oli-Ar committed:
Ported specialisation keyword to HEAD

Ported specialisation keyword from previous commits to new HEAD, as it
was easier than rebasing, as other components had significantly changed.

CAG2Mark committed:
Miscellaneous Fixes (#280)

Co-authored-by: Anson Yeung <nanolify@gmail.com>

Oli-Ar committed:
Added logic to pass specialisation flags through elaborator

As part of moving specialiser code to the new head rewrote logic to pass
the specialisation flags through the compiler - now has a much smaller
footprint.

Oli-Ar committed:
Brought back specialiser logic

Oli-Ar committed:
Pushed majority of logic for the simple sub algorithm

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Specialiser.scala:
```diff
@@ -0,0 +1,753 @@
+    case Primitive(name: Str)
+    case Function(lhs: SimpleType, rhs: SimpleType)
+    case Record(fields: Ls[(Str, SimpleType)])
+    case ClassType(sym: ClassSymbol, supers: Ls[SimpleType], members: Map[Str, SimpleType])
```
This is more like `ClassInfo` (which is _not_ a `SimpleType`). The types of actual class instances should be a different constructor (which points back to the `ClassInfo`).

LPTK commented:
TODO: Write specification/architecture of algorithm first.

You need to attach the flows of normal parameters to the flows of specialized ones.

Example:

```
fun foo(spec x: ?a, p1: ?p1, p2: ?p2) = ...

fun bar(x: ?b, arg: ?arg) =
  foo(x, arg, 123)
// At this point, we know we'll need to specialize this call to `foo` for each new concrete LB coming into ?b
// Each of these specializations will need to create the flows `?arg <: ?p1` and `123 <: ?p2`.
```

Another way of thinking about this: specialized functions are polymorphic, but the polymorphic instances are indexed not by the individual call sites but by the list of concrete shapes flowing into their specialized parameters.

```
fun foo(spec x, spec y, p1, p2) = ... constraint ?p1 <: ?p2 -> ?result
foo: (x, y) => forall p1, p2, result where { p1 <: p2 -> result }. T
```

In this approach, we remember whatever flows into `(p1, p2)` by attaching it (indexing it) to what flows into `(x, y)`. Only when the `(x, y)` flow becomes concrete do we look up the specialization table: then, if there is not yet an entry, we instantiate the polymorphic type scheme; afterwards, we connect the stashes `(p1, p2)` flows with the corresponding instantiated signature.

Note: Similarly, `ClassInfo`s themselves should be indexed on their spec params.

LPTK commented:
Another interesting thing: the calls to specialized functions happening in non-specialized contexts may have to be turned into pattern matches.

Reusing the previous example:

```fs
fun foo(spec x, p1, p2) = ...

fun bar(x, arg) =
  foo(x, arg, 123)

bar(Int, 0)
bar(Str, true)
```

This should basically specialize into:

```fs
fun foo_Int(x, p1, p2) = ...
fun foo_Str(x, p1, p2) = ...

fun bar(x, arg) =
  if x is
    Int then foo_Int(x, arg, 123)
    Str then foo_Str(x, arg, 123)

bar(Int, 0)
bar(Str, true)
```

In fact, at least conceptually, we should probably always generate pattern matching by default and possibly optimize it later. So even a direct call like

```fs
foo(42, true, 0)
```

should be turned into

```fs
let $arg = 42
if $arg is
  Int then foo_Int($arg, true, 0)
```

which will be optimized the the expected

```fs
let $arg = 42
foo_Int($arg, true, 0)
```

So this looks pretty similar to the earlier [defunctionalization/demethodization project](https://github.com/hkust-taco/mlscript/pull/222)!

Oli-Ar committed:
Pulled class info out from SimpleType to new class

Oli-Ar committed:
Fixed type inference for classes and some types of functions

Oli-Ar committed:
Fixed  relationship

Oli-Ar committed:
Code cleanup and minor fixes

Oli-Ar committed:
Added type inference for  statements

Oli-Ar committed:
Delayed processing of TermDefs fixing some class of inference bugs

Oli-Ar committed:
Specialisation inference

Oli-Ar committed:
Generating specialised functions

Oli-Ar committed:
Arguments to specialised parameters are now extracted

Oli-Ar committed:
Cleaned out tests

Oli-Ar committed:
Specialiser type pattern matching added

Oli-Ar committed:
Specialiser type pattern matching added

Oli-Ar committed:
Renaming

Oli-Ar committed:
Handling specialisation inside specialised functions better

Oli-Ar committed:
Improved specificity of specialiser

Oli-Ar committed:
Added typeof function check support

I'm not sure this is desired behaviour but I added this in case a user
specialises a parameter which takes in a lambda, which currently at
least, requires checking if a value is of type function

Oli-Ar committed:
Small features & Bug Fixes

Added small features: specialisation in specialised functions
being more specific, lambda support
Fixed many bugs including not properly hoisting functions, issues with
application specialisation, issues in the constraining of some types

Oli-Ar commented:
The approach I took to specialisation works by taking the SimpleType that a parameter would be and then wrapping it in a SpecialisableType SimpleType, which stores the underlying SimpleType, along with a SpecPoint instance, which stores more details about the specialisation.

Specifically I store any known concrete types, any variables that flow into it, and any flows out of it into other SpecialisableTypes. With this, I can do one pass of the program, and constrain the types, and while I constrain I fill in the rest of the information about the specialised parameter. Then with all of the SpecialisableTypes I can figure out what each function should be specialised to. Then on the next pass I case use that table to actually specialise the functions and the applications of the functions.

The reason I did it this way over the linked list was just that I was struggling to get the linked list method working, and I had this idea so tried it quickly and it worked better than anything I'd managed with the first approach so I decided to keep at it for a bit longer.

Oli-Ar commented on hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala:
```diff
@@ -307,6 +307,7 @@ class JSBuilder(using TL, State, Ctx) extends CodeBuilder:
           case Elaborator.ctx.builtins.Num => doc"typeof $sd === 'number'"
           case Elaborator.ctx.builtins.Bool => doc"typeof $sd === 'boolean'"
           case Elaborator.ctx.builtins.Int => doc"globalThis.Number.isInteger($sd)"
+          case Elaborator.ctx.builtins.Function => doc"typeof $sd === 'function'"
```
I added this as I had a test where I specialise a parameter which I pass a lambda to, so when I generate the pattern matching for the application I needed to be able to match the lambda, but I'd just like to check that this is okay to do since I'm not very familiar with the codegen side of mlscript.

LPTK commented on hkmc2/shared/src/test/mlscript/specialiser/Basic.mls:
```diff
@@ -200,86 +289,40 @@ foo("Str")
 
-
-:sjs
-fun foo(x, y) =
```
Any reason to remove this test?

Oli-Ar replied:
When debugging, I found it gave the same error as the test with `add`, or the test with `Pair`, but was harder to work with than either of them, so I removed it. I've added it back for consistency in commit c08a954.

LPTK commented on hkmc2DiffTests/src/test/scala/hkmc2/MLsDiffMaker.scala:
```diff
@@ -245,8 +245,8 @@ abstract class MLsDiffMaker extends DiffMaker:
-    val checker = new semantics.SimpleSub(curCtx, stl)
-    val spt = checker.analyzeTermTypes(e).asInstanceOf[semantics.Term.Blk]
+    val checker = new semantics.Specialiser(curCtx, stl)
+    val spt = checker.topLevel(e).asInstanceOf[semantics.Term.Blk]
```
Why is the cast needed here but not in `MLsCompiler`? And if we really need a `Blk`, couldn't you just make `topLevel` take and return one?

Oli-Ar committed:
Re-added removed test

Oli-Ar committed:
Improved specialisation

Improved specialisation by making spec points be application sites
rather than individual parameters. This allows for the flows of types to
be tracked neater through a program and allow for less unnecessary code
to be generated.

In addition added better suppoer for higher level specialisation —
passing specialised functions to other functions — however this is still
buggy.

LPTK commented:
This is unlikely to progress further, so I marked it to revisit later and will close it for now.
